### PR TITLE
Allow user to set SGA and PGA memory

### PR DESCRIPTION
**Pull request by rishabh20**

### CODE REVIEW

**Comment by rishabh20** [comment]
This PR addresses issue #1575

**Review by gvenzl — CHANGES_REQUESTED** [review]
This only sets `SGA_TARGET` and `PGA_AGGREGATE_TARGET` on database creation, but will not work when a new container is started with an existing database in a volume.
Because a user cannot easily tell whether a container just starts a database or creates a new one, especially in a K8s environment, it can lead to confusion by the user when setting these parameters and see that they won't have any effect (because the DB was already in an existing volume).

It would be better if we could add to the startup routine a check that verifies the memory parameters and reset them in case they are not what the environment variables are set to (or just ignored everything in case the environment variables are not set at all).

File: OracleDatabase/SingleInstance/dockerfiles/19.3.0/runOracle.sh [hunk 1]
```diff
@@ -181,7 +181,7 @@ else
   rm -f $ORACLE_HOME/network/admin/tnsnames.ora
    
   # Create database
-  $ORACLE_BASE/$CREATE_DB_FILE $ORACLE_SID $ORACLE_PDB $ORACLE_PWD;
+  $ORACLE_BASE/$CREATE_DB_FILE $ORACLE_SID $ORACLE_PDB $ORACLE_PWD || exit 1;
```
[the remark this answers is not in the document]
Comment (rishabh20) [reply OracleDatabase/SingleInstance/dockerfiles/19.3.0/runOracle.sh:184]: I tested this. Without `|| exit 1`, the control comes back to runOracle.sh and it proceeds further even though createDB.sh issues `exit 1`;

File: OracleDatabase/SingleInstance/README.md [hunk 2]
```diff
@@ -59,6 +59,8 @@ To run your Oracle Database Docker image use the **docker run** command as follo
 	-e ORACLE_SID=<your SID> \
 	-e ORACLE_PDB=<your PDB name> \
 	-e ORACLE_PWD=<your database passwords> \
+	-e SGA_SIZE=<your database sga memory in MB> \
```
[the remark this answers is not in the document]
Comment (rishabh20) [reply OracleDatabase/SingleInstance/README.md:62]: Addressed.

File: OracleDatabase/SingleInstance/README.md [hunk 3]
```diff
@@ -59,6 +59,8 @@ To run your Oracle Database Docker image use the **docker run** command as follo
...
+	-e PGA_SIZE=<your database pga memory in MB> \
```
[the remark this answers is not in the document]
Comment (rishabh20) [reply OracleDatabase/SingleInstance/README.md:63]: Addressed.

**Comment by rishabh20** [comment]
@gvenzl Lata, Yunus, and I had a discussion on honoring these params when datafiles are reused. We also went over this comment of yours: https://github.com/oracle/docker-images/pull/1576#issuecomment-668143137
We cannot retrieve the configured parameter without having to startup the database, retrieve the memory, and then restart the DB with the new value if it has changed. 
At this point, we want to give this functionality as we see a lot of issues in 19c where dbca is calculating wrong memory values since it is agnostic of container memory limits setup by k8s pod specs or other means. To mitigate such issues, the customer can make use of these exposed params.
We were thinking to roll this out and if it is needed, we can honor the exposed params during db startup as a future enhancement. For the time being, I have mentioned in the documentation that these values can be set only for the first time when db is created.

File: OracleDatabase/SingleInstance/dockerfiles/19.3.0/runOracle.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (gvenzl) [reply OracleDatabase/SingleInstance/dockerfiles/19.3.0/runOracle.sh:184]: Ok great, thanks!

**Comment by gvenzl** [comment]
Yeah, makes sense. I still am concerned though that we will remove one confusion for another (`dbca` getting the wrong memory settings vs. "but I told the container to use N MB of memory). A compromise could be to call the parameters perhaps `INIT_SGA_SIZE` and `INIT_PGA_SIZE`, that way even if a user doesn't read the documentation (which is very often these days), it's clear that the parameters are for initial settings only. What do you think about that?

**Comment by rishabh20** [comment]
Yeah, prefixing INIT makes sense. I am making this change and pushing it. If you do not have any other comments/concerns, please merge it after my push.

**Review by gvenzl — APPROVED** [review]

**Comment by piotrekfus91** [comment]
I'm not convinced if this is the right solution. Image sample usage: docker-compose. If you have docker-compose that you commit to repo, you set this to fixed value. Now, if you have different machines - on some of it may work, whereas on others not.

**Comment by noeldcosta** [comment]
@rishabh20 what are the equivalent values for INIT_SGA_SIZE and INIT_PGA_SIZE to set it to the earlier behavior of using 2GB. In our use case the build machine typically has 8/16 OCPUs (VM.Standard2.8/2.16). We then create trial editions of our product that users typically run on VM.Standard2.4/2.2. We'd like to revert to the previous behavior of using 2GB so that it gives us portability to move the DB around.

**Comment by rishabh20** [comment]
@noeldcosta SGA + PGA accounts for total memory which you want to keep at 2GB. The recommended value for SGA is 75% of total memory. So, SGA and PGA would be approx. 1.5GB and 0.5GB respectively.

**Comment by noeldcosta** [comment]
thanks Rishabh

**Comment by felipecrs** [comment]
From what I understood, you can also switch to the previous behavior by NOT setting any of these flags and limiting the number of CPUs to the maximum of eight during database creation. Like: `docker build --cpuset-cpus 0-7`.

**Comment by felipecrs** [comment]
But what I really don't understand is why this count is based on the number of CPUs? What does it have to do with the memory?

**Comment by felipecrs** [comment]
> @noeldcosta SGA + PGA accounts for total memory which you want to keep at 2GB. The recommended value for SGA is 75% of total memory. So, SGA and PGA would be approx. 1.5GB and 0.5GB respectively.

Just for reference, it would be `--env INIT_SGA_SIZE=1536 --env INIT_PGA_SIZE=512`.

**Pull request merged**